### PR TITLE
multistream: Dialer deny multiple /multistream/1.0.0 headers

### PR DESCRIPTION
**Pull request by lexnv**
This PR ensures that the `/multistream/1.0.0` header can only be received once during a protocol negotiation.
This is fixed for both the V1 and V1Lazy variations.

While at it, have added a few tests to check:
- protocol negotiation
- protocol negotiation failure due to unsupported protocols
- lazy protocol negotiation awaits protocol message before closing (to be fixed by: https://github.com/paritytech/litep2p/issues/60)
- low level protocol listener that writes the expected messages
- low level protocol listener that submits multiple headers, both for V1 and V1 lazy

Closes: https://github.com/paritytech/litep2p/issues/59

cc @paritytech/networking

### CODE REVIEW

**Comment by dmitry-markin** [comment]
@lexnv We have a pending PR not yet merged that fixes WebRTC: https://github.com/paritytech/litep2p/pull/51.

I'd double check that we don't have conflicting changes, probably merging that one first.

**Comment by lexnv** [comment]
That makes sense! Thanks for the info 🙏 

I think would proly be good to have the CI up and running before we merge any other PRs (I've also put https://github.com/paritytech/litep2p/pull/57 on hold until we get more confidence in our testing)

After having a quick look over the WebRTC PR, it seems that the changes will not conflict:
- as the main change is done to `DialerState` and adding tests for `DialerState` (prob need to double check names)
- and it does not touch the negotiating part of the code

Yep, I think would be good to get that PR in first to ensure we don't break anything with these changes

File: src/multistream_select/dialer_select.rs
Comment (dmitry-markin) [inline src/multistream_select/dialer_select.rs:426]: nit: Is the type annotation required here and with other `tokio::spawn()` invokations?

File: src/multistream_select/dialer_select.rs
Comment (dmitry-markin) [inline src/multistream_select/dialer_select.rs:556]: nit: May be add a remark that we should change this to `.is_ok()` and update the comment above after merging https://github.com/paritytech/litep2p/pull/62?

**Review by dmitry-markin — APPROVED** [review]
LGTM! Nice use of `futures_ringbuf` in tests, didn't know about it.

File: src/multistream_select/dialer_select.rs
Comment (lexnv) [reply src/multistream_select/dialer_select.rs:426]: Unfortunately, I think so because we only return `Ok` from those futures, and the compiler cannot deduce the `Err` type

Comment (dmitry-markin) [reply src/multistream_select/dialer_select.rs:426]: If we never return `Err`, why do we need to return `Result`?

Comment (lexnv) [reply src/multistream_select/dialer_select.rs:426]: Thanks for spotting this! Don't remember why I've added the `Ok()`, probably wanted to use the `?` operator. I've removed the type annotation 🙏

**Pull request merged**
